xtask/src/commands/release/plan.rs

```diff
@@ -0,0 +1,317 @@
+use clap::Args;
+use esp_metadata::Chip;
+use serde::{Deserialize, Serialize};
+use strum::IntoEnumIterator;
+
+use crate::{
+    Package,
+    cargo::CargoToml,
+    commands::{checker::min_package_update, do_version_bump},
+};
+
+#[derive(Debug, Args)]
+pub struct PlanArgs {
+    /// The packages to be released.
+    #[arg(value_enum, default_values_t = Package::iter())]
+    packages: Vec<Package>,
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize)]
+enum VersionBump {
+    PreRelease(String),
+    Patch,
+    Minor,
+    Major,
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize)]
+struct PackagePlan {
+    package: Package,
+    semver_checked: bool,
+    current_version: semver::Version,
+    new_version: semver::Version,
+    tag_name: String,
+    /// The version bump that will be applied to the package.
+    bump: VersionBump,
+}
+
+/// A release plan is a list of packages and their version increments.
+///
+/// The plan is generated by analyzing the workspace and its dependencies.
+/// The order of the packages in the plan is important, as it determines the
+/// order in which the packages are released.
+#[derive(Debug, Clone, Deserialize, Serialize)]
+struct Plan {
+    packages: Vec<PackagePlan>,
+}
+
+pub fn plan(workspace: &Path, args: PlanArgs) -> Result<()> {
+    // Recursively collect dependencies. A bit inefficient, but we don't need to
+    // sort a lot.
+    let mut packages_to_release = args
+        .packages
+        .iter()
+        .filter(|p| p.is_published(workspace))
+        .flat_map(|p| related_crates(workspace, *p))
+        .collect::<Vec<_>>();
+
+    packages_to_release.sort();
+    packages_to_release.dedup();
+
+    let mut package_tomls = packages_to_release
+        .iter()
+        .map(|pkg| CargoToml::new(workspace, *pkg).map(|cargo_toml| (*pkg, cargo_toml)))
+        .collect::<Result<HashMap<_, _>>>()?;
+
+    // Determine package dependencies (package -> dependencies)
+    let mut dep_graph = HashMap::new();
+    for (package, toml) in package_tomls.iter_mut() {
+        dep_graph.insert(*package, toml.repo_dependencies());
+    }
+
+    // Topological sort the packages into a release order.
+    let sorted = topological_sort(&dep_graph);
+    log::debug!("Sorted packages: {:?}", sorted);
+
+    for package in sorted.iter() {
+        ensure!(
+            package_tomls[&package].is_published(),
+            "Cannot release {package}: package is not published. The command \
+            would have released the following packages: {}",
+            sorted
+                .iter()
+                .map(|p| p.to_string())
+                .collect::<Vec<_>>()
+                .join(", ")
+        );
+    }
+
+    // Gather semver bump requirements
+    let mut update_amounts = vec![];
+    let all_chips = Chip::iter().collect::<Vec<_>>();
+    for package in sorted.iter().copied() {
+        let amount = if package_changed_since_last_release(workspace, package) {
```

Comment:
I'm a bit concerned about this check, might we end up bumping/releasing a package even if we make some minor unrelated change within the package? I.e a small README adjustment would return true here, if I'm understanding correctly?

Reply:
Given that we parse the changelogs, can we add some more logic here to ensure that a relevant change is actually being made?

Reply:
Hmm but maybe the real "solution" here is to semvercheck more packages. I'll keep this review (ramble) here, just in case you have any suggestions.

Reply:
In any case, you'll be able to delete the package from the release plan to prevent it from being released.

Reply:
That's fair, I think we can mark this as a "read the output from the plan to ensure it's correct" situation.